### ISSUE TITLE
Bug: while_loop gives invalid gradients (NaN) even when loss is computed using only non NaNs

### ISSUE THREAD

**Issue by rdipietro** [tensorflow/tensorflow#6209]
### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?

https://github.com/tensorflow/tensorflow/issues/2540 is the only thing that is seemingly related, but it's a different issue.

### Environment info
Operating System: Ubuntu 14.04

Installed version of CUDA and cuDNN: This bug occurs even if I only use CPUs.

Compiled from source. Commit hash: a5074383617a9947f248a0ddd56b94f9fb0f970b

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

Overview: Build a minimalist RNN-like model and compute derivatives with respect to the loss. Allow the ability to specify invalid time steps using NaNs.

Behavior:
- If no NaNs are included, it works
- If only one invalid time step is included, it still works (odd)
- If two or more invalid time steps are included, it fails.
- If we change `INVALID_FLOAT_SYMBOL` to, say, `0.12345`, then it works as expected.

```
import tensorflow as tf
import numpy as np

INVALID_FLOAT_SYMBOL = np.nan

w = tf.Variable(1.0)
inputs = tf.placeholder(tf.float32, shape=[None])
targets = tf.placeholder(tf.float32, shape=[None])
T = tf.size(inputs)

x_ta = tf.TensorArray(tf.float32, size=T).unpack(inputs)
h_ta = tf.TensorArray(tf.float32, size=T)

def cond(t, h, h_ta):
  return tf.less(t, T)

def body(t, h, h_ta):
  x = x_ta.read(t)
  h = w * h + x
  h_ta = h_ta.write(t, h)
  return t + 1, h, h_ta

t = tf.constant(0)
h = tf.constant(0.0)
_, _, h_ta = tf.while_loop(cond, body, [t, h, h_ta])
outputs = h_ta.pack()

invalid_mask = tf.is_nan(inputs) if np.isnan(INVALID_FLOAT_SYMBOL) else tf.equal(inputs, INVALID_FLOAT_SYMBOL)
valid_mask = tf.logical_not(invalid_mask)

valid_outputs = tf.boolean_mask(outputs, valid_mask)
valid_targets = tf.boolean_mask(targets, valid_mask)

valid_losses = tf.nn.sigmoid_cross_entropy_with_logits(valid_outputs, valid_targets)
valid_loss = tf.reduce_sum(valid_losses)

grad, = tf.gradients(valid_loss, w)

sess = tf.Session()
sess.run(tf.initialize_all_variables())
```

### Logs or other output that would be helpful
(If logs are large, please upload as attachment or provide link).

```
# Case 1: INVALID_FLOAT_SYMBOL = np.nan, but no INVALID symbol is used.
print(sess.run(grad, feed_dict={inputs: [1.0, 2.0, 3.0],
                                targets: [1.0, 1.0, 1.0]}))
# Correctly prints -0.0573164

# Case 2: INVALID_FLOAT_SYMBOL = np.nan, and one INVALID symbol is used.
print(sess.run(grad, feed_dict={inputs: [1.0, 2.0, 3.0, INVALID_FLOAT_SYMBOL],
                                targets: [1.0, 1.0, 1.0, INVALID_FLOAT_SYMBOL]}))
# Correctly prints -0.0573164

# Case 3: INVALID_FLOAT_SYMBOL = np.nan, and two INVALID symbols are used.
print(sess.run(grad, feed_dict={inputs: [1.0, 2.0, 3.0, INVALID_FLOAT_SYMBOL, INVALID_FLOAT_SYMBOL],
                                targets: [1.0, 1.0, 1.0, INVALID_FLOAT_SYMBOL, INVALID_FLOAT_SYMBOL]}))
# Incorrectly prints nan

# Case 4: INVALID_FLOAT_SYMBOL = 0.12345, and two INVALID symbols are used.
print(sess.run(grad, feed_dict={inputs: [1.0, 2.0, 3.0, INVALID_FLOAT_SYMBOL, INVALID_FLOAT_SYMBOL],
                                targets: [1.0, 1.0, 1.0, INVALID_FLOAT_SYMBOL, INVALID_FLOAT_SYMBOL]}))
# Correctly prints -0.0573164
```

**Comment by rdipietro** [comment]
Temporary fix for anyone else with custom RNN implementations with the desire to indicate specific invalid time steps instead using the `sequence_length` option provided in TensorFlow. (This makes writing various models, e.g. time-step-level regression, sequence-level regression, time-step-level classification, sequence-level classification, or even regression / classification for scattered time steps that aren't just 'all' or 'at the end'.)

If we pad the inputs with 0.0 instead of NaN, this bug will never occur. We can then still pad targets with NaNs and mask using the targets.

**Comment by michaelisard** [comment]
@yuanbyu I suspect the issue is because of the eager processing of conditionals: can you comment?

**Comment by yuanbyu** [comment]
The problem is that backprop uses tensors from forward computation, and some of those tensors are nan because the nans in `inputs`.  Why don't you mask `inputs` instead

`valid_inputs = tf.boolean_mask(inputs, valid_mask)
`
?

**Comment by rdipietro** [comment]
I don't see why the gradient with respect to time step 5 (for example) should depend in any way on the information at time step 10. Is this actually happening?

The logs I include above suggest that this isn't happening: it works if there is only one NaN after the valid data.

Your suggested workaround to the bug works for this simplified case but doesn't work in the general case. In a real RNN setting, the batch is made up of multiple sequences, each with NaNs at different places, so flattening along batch/time with `tf.boolean_mask` does not work. (We could possibly post process segments independently, but this would a) lead to messy code and b) make us lose the efficiency gains from batch processing.)

**Comment by yuanbyu** [comment]
The problem is about the nan steps. In the backprop, after boolean_mask, zeros are used as gradients to continue the back propagation. It could get to an op whose gradient uses some nans produced in the forward.

**Comment by rdipietro** [comment]
Hmm. To me it seems like it's a bug, especially because it doesn't lead to NaNs if exactly one NaN input is included after the valid inputs. But in any case if it's deemed too awkward to fix then it's not too big of a deal; since I encountered this I've been marking invalid time steps with 0s.

**Comment by yuanbyu** [comment]
I would consider it not a bug. So I will close it for now.
